File: plug-ins/Apple.CoreHaptics/Native/CoreHapticsWrapper/Haptics/HapticEngine/CHHapticEngine.swift
//
//  CHHapticEngine.swift
//  CoreHapticsWrapper
//
//  Copyright © 2021 Apple, Inc. All rights reserved.
//
import Foundation
import CoreHaptics

public typealias StoppedWithReasonCallback = @convention(c) (UnsafeRawPointer, Int32) -> Void;
public typealias AllPlayersFinishedCallback = @convention(c) (UnsafeRawPointer) -> Int32;

fileprivate var _implementation : CHHapticEngine_Protocol? = nil;

fileprivate func _instance() -> CHHapticEngine_Protocol {
    if _implementation == nil {
        if #available(iOS 13, tvOS 14.0, macOS 11, *) {
            _implementation = CHHapticEngine_Supported();
        } else {
            _implementation = CHHapticEngine_Unsupported();
        }
    }
    
    return _implementation!;
}

@_cdecl("CoreHaptics_CHHapticEngine_Create")
public func CoreHaptics_CHHapticEngine_Create
(
    onError: ErrorCallback,
    onStopped: @escaping StoppedWithReasonCallback,
    onReset: @escaping SuccessWithPointerCallback
) -> UnsafeMutableRawPointer?
{
    return _instance().Create(onError: onError, onStopped: onStopped, onReset: onReset);
}

@_cdecl("CoreHaptics_CHHapticEngine_Destroy")
public func CoreHaptics_CHHapticEngine_Destroy
(
    enginePtr: UnsafeRawPointer
)
{
    _instance().Destroy(enginePtr: enginePtr);
}

@_cdecl("CoreHaptics_CHHapticEngine_Start")
public func CoreHaptics_CHHapticEngine_Start
(
    enginePtr: UnsafeRawPointer,
    onError: @escaping ErrorWithPointerCallback
)
{
    _instance().Start(enginePtr: enginePtr, onError: onError);
}

@_cdecl("CoreHaptics_CHHapticEngine_Stop")
public func CoreHaptics_CHHapticEngine_Stop
(
    enginePtr: UnsafeRawPointer,
    onError : @escaping ErrorWithPointerCallback
)
{
    _instance().Stop(enginePtr: enginePtr, onError: onError);
}

@_cdecl("CoreHaptics_CHHapticEngine_PlayPatternFromUrl")
public func CoreHaptics_CHHapticEngine_PlayPatternFromUrl
(
    enginePtr: UnsafeRawPointer,
    ahapUrl: char_p,
    onError: @escaping ErrorWithPointerCallback
)
{
    _instance().PlayPatternFromUrl(enginePtr: enginePtr, ahapUrl: ahapUrl, onError: onError);
}

@_cdecl("CoreHaptics_CHHapticEngine_PlayPatternFromJson")
public func CoreHaptics_CHHapticEngine_PlayPatternFromJson
(
    enginePtr: UnsafeRawPointer,
    ahapJson: char_p,
    onError: @escaping ErrorWithPointerCallback
)
{
    _instance().PlayPatternFromJson(enginePtr: enginePtr, ahapJson: ahapJson, onError: onError);
}

@_cdecl("CoreHaptics_CHHapticEngine_MakePlayer")
public func CoreHaptics_CHHapticEngine_MakePlayer
(
    enginePtr: UnsafeRawPointer,
    ahapJson: char_p,
    onError: ErrorCallback
) -> UnsafeMutableRawPointer?
{
    return _instance().MakePlayer(enginePtr: enginePtr, ahapJson: ahapJson, onError: onError);
}

@_cdecl("CoreHaptics_CHHapticEngine_MakeAdvancedPlayer")
public func CoreHaptics_CHHapticEngine_MakeAdvancedPlayer
(
    enginePtr: UnsafeRawPointer,
    ahapJson: char_p,
    onFinishedPlaying: @escaping SuccessWithPointerCallback,
    onError: ErrorCallback
) -> UnsafeMutableRawPointer?
{
    return _instance().MakeAdvancedPlayer(enginePtr: enginePtr, ahapJson: ahapJson, onFinishedPlaying: onFinishedPlaying, onError: onError);
}

@_cdecl("CoreHaptics_CHHapticEngine_NotifyWhenPlayersFinished")
public func CoreHaptics_CHHapticEngine_NotifyWhenPlayersFinished
(
    enginePtr: UnsafeRawPointer, leaveEngineRunning: Bool,
    onAllPlayersFinished: @escaping AllPlayersFinishedCallback,
    onError: @escaping ErrorWithPointerCallback
)
{
    _instance().NotifyWhenPlayersFinished(enginePtr: enginePtr, leaveEngineRunning: leaveEngineRunning, onAllPlayersFinished: onAllPlayersFinished, onError: onError);
}

@_cdecl("CoreHaptics_CHHapticEngine_HardwareSupportsHaptics")
public func CoreHaptics_CHHapticEngine_HardwareSupportsHaptics() -> Bool
{
    return _instance().HardwareSupportsHaptics();
}

@_cdecl("CoreHaptics_CHHapticEngine_Get_IsAutoShutdownEnabled")
public func CoreHaptics_CHHapticEngine_Get_IsAutoShutdownEnabled
(
    enginePtr: UnsafeRawPointer
) -> Bool
{
    return _instance().Get_IsAutoShutdownEnabled(enginePtr: enginePtr);
}

@_cdecl("CoreHaptics_CHHapticEngine_Set_IsAutoShutdownEnabled")
public func CoreHaptics_CHHapticEngine_Set_IsAutoShutdownEnabled
(
    enginePtr: UnsafeRawPointer,
    isAutoShutdownEnabled: Bool
)
{
    _instance().Set_IsAutoShutdownEnabled(enginePtr: enginePtr, isAutoShutdownEnabled: isAutoShutdownEnabled);
}

@_cdecl("CoreHaptics_CHHapticEngine_Get_PlaysHapticsOnly")
public func CoreHaptics_CHHapticEngine_Get_PlaysHapticsOnly
(
    enginePtr: UnsafeRawPointer
) -> Bool
{
    return _instance().Get_PlaysHapticsOnly(enginePtr: enginePtr);
}

@_cdecl("CoreHaptics_CHHapticEngine_Set_PlaysHapticsOnly")
public func CoreHaptics_CHHapticEngine_Set_PlaysHapticsOnly
(
    enginePtr: UnsafeRawPointer,
    playsHapticsOnly: Bool
)
{
    _instance().Set_PlaysHapticsOnly(enginePtr: enginePtr, playsHapticsOnly: playsHapticsOnly);
}

@_cdecl("CoreHaptics_CHHapticEngine_Get_IsMutedForAudio")
public func CoreHaptics_CHHapticEngine_Get_IsMutedForAudio
(
    enginePtr: UnsafeRawPointer
) -> Bool
{
    return _instance().Get_IsMutedForAudio(enginePtr: enginePtr);
}

@_cdecl("CoreHaptics_CHHapticEngine_Set_IsMutedForAudio")
public func CoreHaptics_CHHapticEngine_Set_IsMutedForAudio
(
    enginePtr: UnsafeRawPointer,
    isMutedForAudio: Bool
)
{
    _instance().Set_IsMutedForAudio(enginePtr: enginePtr, isMutedForAudio: isMutedForAudio);
}

@_cdecl("CoreHaptics_CHHapticEngine_Get_IsMutedForHaptics")
public func CoreHaptics_CHHapticEngine_Get_IsMutedForHaptics
(
    enginePtr: UnsafeRawPointer
) -> Bool
{
    return _instance().Get_IsMutedForHaptics(enginePtr: enginePtr);
}

@_cdecl("CoreHaptics_CHHapticEngine_Set_IsMutedForHaptics")
public func CoreHaptics_CHHapticEngine_Set_IsMutedForHaptics
(
    enginePtr: UnsafeRawPointer,
    isMutedForHaptics: Bool
)
{
    _instance().Set_IsMutedForHaptics(enginePtr: enginePtr, isMutedForHaptics: isMutedForHaptics);
}

@_cdecl("CoreHaptics_CHHapticEngine_Get_CurrentTime")
public func CoreHaptics_CHHapticEngine_Get_CurrentTime
(
    enginePtr: UnsafeRawPointer
) -> Double
{
    return _instance().Get_CurrentTime(enginePtr: enginePtr);
}

File: plug-ins/Apple.CoreHaptics/Apple.CoreHaptics_Unity/Assets/Apple.CoreHaptics/Source/CHHapticEngine/CHHapticEngine.cs
using System;
using System.Collections.Generic;
using System.Linq;
using AOT;
using UnityEngine;

namespace Apple.CoreHaptics
{
	public class CHHapticEngine
	{
		#region Delegates
		/// <summary>
		/// The delegate signature that a developer should use for subscribing
		/// to the Reset Event.
		/// See <see cref="http://developer.apple.com/documentation/corehaptics/chhapticengine/3043661-resethandler"/>
		/// </summary>
		public delegate void EngineResetHandler();

		/// <summary>
		/// The delegate signature that a developer should use for subscribing
		/// to the EngineStoppedHandler Event.
		/// <see cref="http://developer.apple.com/documentation/corehaptics/chhapticengine/3043665-stoppedhandler"/>
		/// </summary>
		/// <remarks>
		/// The stopped handler is only called when the engine is stopped for a
		/// reason other than a call to CHHapticEngine.StopEngine()
		/// </remarks>
		/// <param name="reason">The reason the Engine stopped</param>
		public delegate void EngineStoppedHandler(CHHapticEngineStoppedReason reason);

		internal delegate void AllPlayersFinishedStatic(IntPtr engPtr);
		public delegate void NotifyAllPlayersFinished();
		#endregion

		#region Static Properties & Events
		private static readonly Dictionary<IntPtr, CHHapticEngine> _pointerToEngines = new Dictionary<IntPtr, CHHapticEngine>();
		private static readonly object _pointerToEnginesLock = new object();

		private static CHException _lastCreationException;
		private static CHHapticEngine _oneShotEngine;

		public static CHHapticEngine OneShotEngine => GetOneShotEngine();

		#endregion

		#region Events
		public event EngineResetHandler Reset;
		public event EngineStoppedHandler Stopped;
		public event NotifyAllPlayersFinished PlayersFinished;
		#endregion

		#region Instance Fields & Properties
		public IntPtr EnginePtr = IntPtr.Zero;

		private readonly List<CHHapticPlayerConfiguration> _playerConfigurations = new List<CHHapticPlayerConfiguration>();
		private CHException _lastOperationException;

		#endregion

		#region Interop Properties
		public bool PlaysHapticsOnly
		{
			get => EnginePtr != IntPtr.Zero && CHHapticEngineInterface.Get_PlaysHapticsOnly(EnginePtr);
			set
			{
				if (EnginePtr != IntPtr.Zero) CHHapticEngineInterface.Set_PlaysHapticsOnly(EnginePtr, value);
			}
		}

		public bool IsMutedForAudio
		{
			get => EnginePtr != IntPtr.Zero && CHHapticEngineInterface.Get_IsMutedForAudio(EnginePtr);
			set
			{
				if (EnginePtr != IntPtr.Zero) CHHapticEngineInterface.Set_IsMutedForAudio(EnginePtr, value);
			}
		}

		public bool IsMutedForHaptics
		{
			get => EnginePtr != IntPtr.Zero && CHHapticEngineInterface.Get_IsMutedForHaptics(EnginePtr);
			set
			{
				if (EnginePtr != IntPtr.Zero) CHHapticEngineInterface.Set_IsMutedForHaptics(EnginePtr, value);
			}
		}

		public bool IsAutoShutdownEnabled
		{
			get => EnginePtr != IntPtr.Zero && CHHapticEngineInterface.Get_IsAutoShutdownEnabled(EnginePtr);
			set
			{
				if (EnginePtr != IntPtr.Zero) CHHapticEngineInterface.Set_IsAutoShutdownEnabled(EnginePtr, value);
			}
		}

		public double CurrentTime => EnginePtr != IntPtr.Zero ? CHHapticEngineInterface.Get_CurrentTime(EnginePtr) : 0;

		public static bool HardwareSupportsHaptics()
		{
			return CHHapticEngineInterface.HardwareSupportsHaptics();
		}
		#endregion

		public CHHapticEngine()
		{
			CreateEngine();
			PostCreateSetup();
		}

		public CHHapticEngine(IntPtr existingEnginePtr)
		{
			EnginePtr = existingEnginePtr;
			PostCreateSetup();
		}

		private void CreateEngine()
		{
			_lastCreationException = null;

			EnginePtr = CHHapticEngineInterface.Create(OnCreateError, OnStoppedWithReason, OnReset);

			if (!(_lastCreationException is null))
				throw _lastCreationException;
		}

		private void PostCreateSetup()
		{
			if (EnginePtr == IntPtr.Zero)
				return;

			lock (_pointerToEnginesLock)
			{
				_pointerToEngines[EnginePtr] = this;
			}
		}

		public void Destroy()
		{
			DestroyAllPlayers();

			if (EnginePtr == IntPtr.Zero)
				return;

			lock (_pointerToEnginesLock)
			{
				_pointerToEngines.Remove(EnginePtr);
			}

			CHHapticEngineInterface.Destroy(EnginePtr);
			EnginePtr = IntPtr.Zero;
		}

		public void Start()
		{
			ResetLastOperationException();
			CHHapticEngineInterface.Start(EnginePtr, OnThrowErrorWithPointer);
			CheckAndThrowLastOperationException();
		}
		public void Stop()
		{
			ResetLastOperationException();
			CHHapticEngineInterface.Stop(EnginePtr, OnThrowErrorWithPointer);
			CheckAndThrowLastOperationException();
		}

		#region Destroy Players
		public void DestroyAllPlayers()
		{
			foreach (var config in _playerConfigurations)
			{
				config.Player?.Destroy();
			}

			_playerConfigurations.Clear();
		}

		public void DestroyPlayer(CHHapticPatternPlayer player)
		{
			var config = _playerConfigurations.FirstOrDefault(c => c.Player == player);
			_playerConfigurations.Remove(config);

			player.Destroy();
		}
		#endregion

		#region Make Players
		public CHHapticPatternPlayer MakePlayer(List<CHHapticEvent> events)
		{
			var ahap = new CHHapticPattern(events);
			return MakePlayer(ahap);
		}

		public CHHapticPatternPlayer MakePlayer(CHHapticPattern ahap)
		{
			var ahapJson = CHSerializer.SerializeForRuntime(ahap);

			ResetLastOperationException();
			var playerId = CHHapticEngineInterface.MakePlayer(EnginePtr, ahapJson, OnThrowErrorWithPointer);
			CheckAndThrowLastOperationException();

			var player = new CHHapticPatternPlayer(playerId, this);
			_playerConfigurations.Add(new CHHapticPlayerConfiguration(ahap, player));
			player.Setup();

			return player;
		}
		#endregion

		#region MakeAdvancdPlayers
		public CHHapticAdvancedPatternPlayer MakeAdvancedPlayer(List<CHHapticEvent> events)
		{
			var ahap = new CHHapticPattern(events);
			return MakeAdvancedPlayer(ahap);
		}

		public CHHapticAdvancedPatternPlayer MakeAdvancedPlayer(CHHapticPattern ahap)
		{
			var ahapJson = CHSerializer.SerializeForRuntime(ahap);

			ResetLastOperationException();
			var playerId = CHHapticEngineInterface.MakeAdvancedPlayer(EnginePtr, ahapJson, CHHapticAdvancedPatternPlayer.OnAdvancedPlayerFinishedPlaying, OnThrowErrorWithPointer);
			CheckAndThrowLastOperationException();

			var player = new CHHapticAdvancedPatternPlayer(playerId, this);
			_playerConfigurations.Add(new CHHapticPlayerConfiguration(ahap, player));
			player.Setup();

			return player;
		}
		#endregion

		#region TryGetPlayerForPattern && TryGetAdvancedPlayerForPattner
		public bool TryGetPlayerForPattern(CHHapticPattern ahap, out CHHapticPatternPlayer player)
		{
			var config = _playerConfigurations.FirstOrDefault(c => c.Ahap == ahap);

			if (config?.Player != null)
			{
				player = config.Player;
				return true;
			}

			player = null;
			return false;
		}

		public bool TryGetAdvancedPlayerForPattern(CHHapticPattern ahap, out CHHapticAdvancedPatternPlayer player)
		{
			var config = _playerConfigurations.FirstOrDefault(c => c.Ahap == ahap);

			if (config?.Player is CHHapticAdvancedPatternPlayer advancedPatternPlayer)
			{
				player = advancedPatternPlayer;
				return true;
			}

			player = null;
			return false;
		}
		#endregion

		#region Refresh Players

		public void RefreshPlayers()
		{
			foreach (var config in _playerConfigurations)
			{
				// Ensure we destroy the player if not already done...
				config.Player.Destroy();

				// Recreate the player...
				var ahapJson = CHSerializer.SerializeForRuntime(config.Ahap);
				var isAdvancedPlayer = config.Player is CHHapticAdvancedPatternPlayer;

				if (isAdvancedPlayer)
				{
					ResetLastOperationException();
					config.Player.PlayerId = CHHapticEngineInterface.MakeAdvancedPlayer(EnginePtr, ahapJson, CHHapticAdvancedPatternPlayer.OnAdvancedPlayerFinishedPlaying, OnThrowErrorWithPointer);
					CheckAndThrowLastOperationException();
				}
				else
				{
					ResetLastOperationException();
					config.Player.PlayerId = CHHapticEngineInterface.MakePlayer(EnginePtr, ahapJson, OnThrowErrorWithPointer);
					CheckAndThrowLastOperationException();
				}
			}
		}
		#endregion

		#region NotifyWhenPlayersFinished

		public void NotifyWhenPlayersFinished(bool leaveEngineRunning = true)
		{
			Debug.Log("Registering notify finished handler");
			CHHapticEngineInterface.NotifyWhenPlayersFinished(EnginePtr, leaveEngineRunning, OnAllPlayersFinished, OnThrowErrorWithPointer);
		}
		#endregion

		#region PlayPatternFromUrl, PlayPatternFromJson, and PlayPatternFromAHAP
		public void PlayPatternFromUrl(string ahapFilePath)
		{
			if (!ahapFilePath.Contains("StreamingAssets") && !ahapFilePath.Contains(Application.streamingAssetsPath))
			{
				throw new ArgumentException("AHAP URL must be in StreamingAssets for proper playback.");
			}
			ResetLastOperationException();
			CHHapticEngineInterface.PlayPatternFromUrl(EnginePtr, ahapFilePath, OnThrowErrorWithPointer);
			CheckAndThrowLastOperationException();
		}

		public void PlayPatternFromJson(string ahapJson)
		{
			ResetLastOperationException();
			CHHapticEngineInterface.PlayPatternFromJson(EnginePtr, ahapJson, OnThrowErrorWithPointer);
			CheckAndThrowLastOperationException();
		}

		public void PlayPattern(CHHapticPattern ahap)
		{
			if (ahap is null)
			{
				throw new ArgumentException("Cannot play a null AHAP.");
			}

			var ahapJson = CHSerializer.SerializeForRuntime(ahap);

			ResetLastOperationException();
			CHHapticEngineInterface.PlayPatternFromJson(EnginePtr, ahapJson, OnThrowErrorWithPointer);
			CheckAndThrowLastOperationException();
		}

		public void PlayPatternFromAhap(TextAsset ahap)
		{
			if (ahap is null)
			{
				throw new ArgumentException("Cannot play a null AHAP.");
			}
			PlayPattern(new CHHapticPattern(ahap));
		}

		public void PlayPatternFromAhap(AHAPAsset ahap)
		{
			if (ahap is null)
			{
				throw new ArgumentException("Cannot play a null AHAP.");
			}

			PlayPattern(ahap.GetPattern());
		}
		#endregion

		#region Protected Utility

		private void ResetLastOperationException()
		{
			_lastOperationException = null;
		}

		private void CheckAndThrowLastOperationException()
		{
			// Throw exception from last native operation...
			if (!(_lastOperationException is null))
			{
				throw _lastOperationException;
			}
		}

		private static CHHapticEngine GetOneShotEngine()
		{
			return _oneShotEngine ?? (_oneShotEngine = new CHHapticEngine());
		}
		#endregion

		#region Static Callbacks

		[MonoPInvokeCallback(typeof(AllPlayersFinishedStatic))]
		private static void OnAllPlayersFinished(IntPtr pointer)
		{
			lock (_pointerToEnginesLock)
			{
				if (_pointerToEngines.ContainsKey(pointer) && !(_pointerToEngines[pointer] is null) && !(_pointerToEngines[pointer].PlayersFinished is null))
				{
					_pointerToEngines[pointer].PlayersFinished?.Invoke();
				}
			}
		}

		[MonoPInvokeCallback(typeof(GenericCallback<IntPtr, CHHapticEngineStoppedReason>))]
		private static void OnStoppedWithReason(IntPtr pointer, CHHapticEngineStoppedReason reason)
		{
			lock (_pointerToEnginesLock)
			{
				if (_pointerToEngines.ContainsKey(pointer) && !(_pointerToEngines[pointer] is null))
				{
					_pointerToEngines[pointer].Stopped?.Invoke(reason);
				}
			}
		}

		[MonoPInvokeCallback(typeof(SuccessWithPointerCallback))]
		private static void OnReset(IntPtr pointer)
		{
			lock (_pointerToEnginesLock)
			{
				if (_pointerToEngines.ContainsKey(pointer) && !(_pointerToEngines[pointer] is null))
				{
					_pointerToEngines[pointer].Reset?.Invoke();
				}
			}
		}

		[MonoPInvokeCallback(typeof(ErrorCallback))]
		private static void OnCreateError(CHError error)
		{
			_lastCreationException = new CHException(error);
		}

		[MonoPInvokeCallback(typeof(ErrorWithPointerCallback))]
		private static void OnThrowErrorWithPointer(IntPtr pointer, CHError error)
		{
			lock (_pointerToEnginesLock)
			{
				if (_pointerToEngines.ContainsKey(pointer) && !(_pointerToEngines[pointer] is null))
				{
					_pointerToEngines[pointer]._lastOperationException = new CHException(error);
				}
			}
		}
		#endregion
	}
}
